File: Hitro_move/src/calculate_tf.cpp
#include <ros/ros.h>
#include <tf/transform_listener.h>

int main(int argc, char** argv)
{
  ros::init(argc, argv, "tf_example");

  ros::NodeHandle nh;

  tf::TransformListener listener;

  std::string target_frame_id = "base_link";//origin
  std::string source_frame_id = "hand_roll_1";//target

  tf::StampedTransform transform;
  try
  {
    listener.waitForTransform(target_frame_id, source_frame_id, ros::Time(0), ros::Duration(3.0));
    listener.lookupTransform(target_frame_id, source_frame_id, ros::Time(0), transform);
    ROS_INFO("Translation: (%f, %f, %f)", transform.getOrigin().x(), transform.getOrigin().y(), transform.getOrigin().z());
    ROS_INFO("Rotation: (%f, %f, %f, %f)", transform.getRotation().x(), transform.getRotation().y(), transform.getRotation().z(), transform.getRotation().w());
  }
  catch (tf::TransformException& ex)
  {
    ROS_ERROR("Failed to lookup transform: %s", ex.what());
  }

  return 0;
}

File: Hitro_move/scripts/calculate_tf.py
import rospy
import tf
from geometry_msgs.msg import Point, Quaternion, TransformStamped

rospy.init_node('tf_example')

# 例として、新しいトランスフォーメーションを計算するためのデータを作成します
translation = Point(x=1.0, y=2.0, z=3.0)
rotation = Quaternion(x=0.0, y=0.0, z=0.0, w=1.0)
child_frame_id = 'base_link'
parent_frame_id = 'hand_roll_1'

# 新しいトランスフォーメーションを作成します
new_transform = TransformStamped()
new_transform.header.stamp = rospy.Time.now()
new_transform.header.frame_id = parent_frame_id
new_transform.child_frame_id = child_frame_id
new_transform.transform.translation = translation
new_transform.transform.rotation = rotation

# トランスフォーメーションを計算します
listener = tf.TransformerROS()
listener.setTransform(new_transform)

# 特定のフレーム間のトランスフォーメーションを取得します
target_frame_id = 'hand_roll_1'
source_frame_id = 'base_link'
try:
    (trans, rot) = listener.lookupTransform(target_frame_id, source_frame_id, rospy.Time(0))
    print("Translation:", trans)
    print("Rotation:", rot)
except (tf.LookupException, tf.ConnectivityException, tf.ExtrapolationException):
    print("Failed to lookup transform.")
